### ISSUE TITLE
Python script to download documents that are using blurred images

### ISSUE THREAD

**Issue by landoncrabtree** [isanchop/stuhack#46]
Some documents implement blurred images (`https://d3tvd1u91rr79.cloudfront.net/a2aa9a4720ce8e5692f963a70e3cfcc9/html/pages/blurred/page3.webp...`). This can be bypassed simply by removing the `/blurred/` from the Cloudfront URL. 

This means that the StuHack extension does not work for these documents, since StuHack currently relies on client-side blurring with the CSS style `filter: blur(2px)` and simply undoing that blurring. I wrote a Python script that when given a URL, it will save the document as a PDF by identifying the Cloudfront CDN URL and constructing a non-blurred URL path. 

```python
from typing import List, Tuple

import requests
from PIL import Image
from io import BytesIO
import re

def get_total_pages(document_link: str) -> int:
    r = requests.get(document_link, headers={"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/601.3.9 (KHTML, like Gecko) Version/9.0.2 Safari/601.3.9"})
    regex = r'data-page-no=\\\"([0-9a-f]+)\\\"'
    matches = re.findall(regex, r.text, re.IGNORECASE)
    total_pages = max(int(match, 16) for match in matches)
    return total_pages

def get_image_info(document_link: str) -> Tuple[str, str, str]:
    r = requests.get(document_link, headers={"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/601.3.9 (KHTML, like Gecko) Version/9.0.2 Safari/601.3.9"})
    regex = r'src="https://(?P<subdomain>[\w.-]+)\.cloudfront\.net/(?P<identifier>[\w.-]+)/html/bg1\.png\?Policy=(?P<policy>[^"]+)"'
    matches = re.search(regex, r.text, re.IGNORECASE)
    if matches:
        subdomain = matches.group("subdomain")
        identifier = matches.group("identifier")
        policy = matches.group("policy").replace("&amp;", "&")
        return subdomain, identifier, policy
    else:
        raise ValueError("Image information not found in the document link.")

def download_images(subdomain: str, identifier: str, policy: str, total_pages: int) -> List[Image.Image]:
    images = []
    for i in range(1, total_pages + 1):
        url = f"https://{subdomain}.cloudfront.net/{identifier}/html/pages/page{i}.webp?Policy={policy}"
        r = requests.get(url, allow_redirects=True)
        try:
            images.append(Image.open(BytesIO(r.content)))
        except Exception as e:
            raise ValueError(f"Failed to download page {i}. Looks like this document is using client-side blurring-- try using the StuHack Chrome extension instead.")
        print(f"[~] Downloaded page {i} of {total_pages}")
    return images

def save_document(output_file: str, images: List[Image.Image]) -> None:
    if images:
        images[0].save(output_file, save_all=True, append_images=images[1:])
        print(f"[+] Document saved to {output_file}")
    else:
        raise ValueError("No images found.")

def main() -> None:
    document_link = input("Enter the link to the document: ")
    if "studocu" not in document_link:
        print("Invalid link")
        return

    output_file = input("Enter the output file name (e.g., output.pdf): ")

    try:
        total_pages = get_total_pages(document_link)
        print(f"[+] Identified {total_pages} pages in the document.")
        subdomain, identifier, policy = get_image_info(document_link)
        images = download_images(subdomain, identifier, policy, total_pages)
        save_document(output_file, images)

    except Exception as e:
        print(f"[-] An error occurred: {e}")

if __name__ == "__main__":
    main()
```

Thought I would leave this here for whoever needs it, since I know there are a few issues open regarding the extension not working. 

TL;DR
- Some documents utilize image blurring on the backend rather than the client-side `filter: blur(2px)`
- Try the StuHack extension first. If it doesn't work, the document is probably using blurred images
- If the extension doesn't work, try this Python script

I was thinking of making a PR to help integrate this into StuHack, but my Javascript knowledge is pretty lacking. I also don't know my way around a lot of the Web API stuff and noticed there's some things StuDocu does like if the `filter: blur(2px)` gets removed, it gets added back. Figured some folks smarter than me could implement my logic into the extension to make it support both types of documents.

**Comment by CHasanJ** [comment]
how to use this cloudfare to unblur

**Comment by 10mounted1** [comment]
i got this as an error in the idle  any ideas ?

Enter the output file name (e.g., output.pdf): 123.pdf
[-] An error occurred: max() arg is an empty sequence

**Comment by landoncrabtree** [comment]
> i got this as an error in the idle any ideas ?
> 
> Enter the output file name (e.g., output.pdf): 123.pdf [-] An error occurred: max() arg is an empty sequence

What’s the link you’re trying?

**Comment by Sigrag97** [comment]
How can i use this script?

**Comment by landoncrabtree** [comment]
> How can i use this script?

Install Python3 & pip.
pip install requests
pip install pillow
python3 <filename>.py

**Comment by Destro9000** [comment]
@landoncrabtree tried to run the Code but it doesn't do anything after inputing the link

**Comment by landoncrabtree** [comment]
> @landoncrabtree tried to run the Code but it doesn't do anything after inputing the link The link im trying is this https://www.studocu.com/it/document/universita-politecnica-delle-marche/biochimica-1/biochimica/30130670

Was able to download it on my end. Make sure you have all the necessary dependencies installed. 

`pip install requests Pillow typing` 

If you need help, feel free to reach out on Discord `__landon`. Will see what I can do!

**Comment by landoncrabtree** [comment]
> great script, ty! feature suggestion: OCR to have a text-searchable pdf as output. didn't have time to implement myself yet, sorry :)

I've made a modified version of the script which supports OCR. 

```python
from typing import List, Tuple # pip install typing

import requests # pip install requests
from PIL import Image # pip install pillow
from io import BytesIO
import re
import subprocess
import os
import fitz # pip install PyMuPDF

def get_total_pages(document_link: str) -> int:
    r = requests.get(document_link, headers={"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/601.3.9 (KHTML, like Gecko) Version/9.0.2 Safari/601.3.9"})
    regex = r'data-page-no=\\\"([0-9a-f]+)\\\"'
    matches = re.findall(regex, r.text, re.IGNORECASE)
    total_pages = max(int(match, 16) for match in matches)
    return total_pages

def get_image_info(document_link: str) -> Tuple[str, str, str]:
    r = requests.get(document_link, headers={"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/601.3.9 (KHTML, like Gecko) Version/9.0.2 Safari/601.3.9"})
    regex = r'src="https://(?P<subdomain>[\w.-]+)\.cloudfront\.net/(?P<identifier>[\w.-]+)/html/bg1\.png\?Policy=(?P<policy>[^"]+)"'
    matches = re.search(regex, r.text, re.IGNORECASE)
    if matches:
        subdomain = matches.group("subdomain")
        identifier = matches.group("identifier")
        policy = matches.group("policy").replace("&amp;", "&")
        return subdomain, identifier, policy
    else:
        raise ValueError("Image information not found in the document link.")

def download_images(subdomain: str, identifier: str, policy: str, total_pages: int) -> List[Image.Image]:
    images = []
    for i in range(1, total_pages + 1):
        url = f"https://{subdomain}.cloudfront.net/{identifier}/html/pages/page{i}.webp?Policy={policy}"
        r = requests.get(url, allow_redirects=True)
        try:
            images.append(Image.open(BytesIO(r.content)))
        except Exception as e:
            raise ValueError(f"Failed to download page {i}. Looks like this document is using client-side blurring-- try using the StuHack Chrome extension instead.")
        print(f"[~] Downloaded page {i} of {total_pages}")
    return images

def save_document(output_file: str, images: List[Image.Image]) -> None:
    
    if images:
        result = fitz.open()
        for i, image in enumerate(images):
            try:
                image.save(f"page{i}.png")
                cmd = f"tesseract page{i}.png page{i} PDF"
                stdout, stderr = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE).communicate()
                # if stderr: 
                #     print(stderr.decode())
                
                with fitz.open(f"page{i}.pdf") as pdf_file:
                    result.insert_pdf(pdf_file)
            
            except Exception as e:
                raise ValueError(f"Failed to save page {i}. {e}")
            finally:
                os.remove(f"page{i}.png")
                os.remove(f"page{i}.pdf")
        result.save(output_file)  
    else:
        raise ValueError("No images found.")

def main() -> None:
    document_link = input("Enter the link to the document: ")
    if "studocu" not in document_link:
        print("Invalid link")
        return

    output_file = input("Enter the output file name (e.g., output.pdf): ")

    try:
        total_pages = get_total_pages(document_link)
        print(f"[+] Identified {total_pages} pages in the document.")
        subdomain, identifier, policy = get_image_info(document_link)
        images = download_images(subdomain, identifier, policy, total_pages)
        save_document(output_file, images)

    except Exception as e:
        print(f"[-] An error occurred: {e}")

if __name__ == "__main__":
    main()
```

You'll need to install PyMuPDF with `pip install PyMuPDF` :)
This will take a pretty long time, since it runs Tessaract on each image. And on that note, you'll need Tessaract installed and in PATH.

**Comment by Destro9000** [comment]
@landoncrabtree i didn't get typing too , i'll try later and see if it will work

**Comment by zN3utr4l** [comment]
@landoncrabtree 
It does not work
> 
> I've made a modified version of the script which supports OCR.
> 

Enter the link to the document: https://www.studocu.com/it/document/universita-degli-studi-di-milano-bicocca/didattica-e-pedagogia-dellinclusione/riassunto-fuori-dai-giochi-la-psicologia-di-fronte-allesclusione-sociale/651768
Enter the output file name (e.g., output.pdf): test.pdf
[+] Identified 14 pages in the document.
[~] Downloaded page 1 of 14
[~] Downloaded page 2 of 14
[~] Downloaded page 3 of 14
[~] Downloaded page 4 of 14
[~] Downloaded page 5 of 14
[~] Downloaded page 6 of 14
[~] Downloaded page 7 of 14
[~] Downloaded page 8 of 14
[~] Downloaded page 9 of 14
[~] Downloaded page 10 of 14
[~] Downloaded page 11 of 14
[~] Downloaded page 12 of 14
[~] Downloaded page 13 of 14
[~] Downloaded page 14 of 14
[-] An error occurred: [WinError 2] Impossibile trovare il file specificato: 'page0.pdf'

**Comment by landoncrabtree** [comment]
> @landoncrabtree It does not work
> 
> > I've made a modified version of the script which supports OCR.
> 
> Enter the link to the document: https://www.studocu.com/it/document/universita-degli-studi-di-milano-bicocca/didattica-e-pedagogia-dellinclusione/riassunto-fuori-dai-giochi-la-psicologia-di-fronte-allesclusione-sociale/651768 Enter the output file name (e.g., output.pdf): test.pdf [+] Identified 14 pages in the document. [~] Downloaded page 1 of 14 [~] Downloaded page 2 of 14 [~] Downloaded page 3 of 14 [~] Downloaded page 4 of 14 [~] Downloaded page 5 of 14 [~] Downloaded page 6 of 14 [~] Downloaded page 7 of 14 [~] Downloaded page 8 of 14 [~] Downloaded page 9 of 14 [~] Downloaded page 10 of 14 [~] Downloaded page 11 of 14 [~] Downloaded page 12 of 14 [~] Downloaded page 13 of 14 [~] Downloaded page 14 of 14 [-] An error occurred: [WinError 2] Impossibile trovare il file specificato: 'page0.pdf'
> 
> But this works:
> 
> [#46 (comment)](https://github.com/isanchop/stuhack/issues/46#issue-2007162161)

You probably don’t have Tesseract installed, since it looks like the conversion from png -> pdf failed. Make sure to install Tesseract and have it in your PATH (ie, running “tesseract” from cmd.exe will show the help message).

**Comment by GemSeller** [comment]
I am getting this error every time:

[~] Downloaded page 4 of 12
[~] Downloaded page 5 of 12
[-] An error occurred: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))

Is there some additional thing I must do?

Thank you!

**Comment by landoncrabtree** [comment]
> I am getting this error every time:
> 
> [~] Downloaded page 4 of 12 [~] Downloaded page 5 of 12 [-] An error occurred: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))
> 
> Is there some additional thing I must do?
> 
> Thank you!

Drop the document link and I’ll investigate. Looks like theres a connection error when trying to retrieve the unblurred page.

**Comment by Fritz914** [comment]
@landoncrabtree Hey I've sent u a request on dc I need ur help. would be pleased if u could help me out
